File: csharp/src/SpringNBTLibrary/Nbt/NbtBinaryWriter.cs
using System.Buffers.Binary;

namespace SpringNBTLibrary.Nbt;

/// <summary>
/// NBT を展開済みのバイト列へ書き出す
/// </summary>
/// <remarks>
/// <para>
/// 出力は一意でなければならない（ラウンドトリップ検証が成立するため）
/// Compound は挿入順のまま、浮動小数点はビットパターンのまま書き出す
/// </para>
/// <para>仕様: <c>docs/spec/10-nbt-binary.md</c> 5章</para>
/// </remarks>
internal sealed class NbtBinaryWriter
{
    private readonly List<byte> buffer = new List<byte>();

    /// <summary>ルートタグを書き出し、結果のバイト列を返す</summary>
    internal byte[] WriteRoot(NamedTag named, NbtFormat format)
    {
        buffer.Add((byte)TagType.Compound);

        if (format == NbtFormat.Java)
        {
            // ファイル形式のルートには名前が付く
            WriteString(named.Name);
        }

        WriteCompoundPayload(named.Tag);
        return buffer.ToArray();
    }

    private void WritePayload(NbtTag tag)
    {
        switch (tag)
        {
            case NbtByte value:
                buffer.Add((byte)value.Value);
                break;
            case NbtShort value:
                WriteInt16(value.Value);
                break;
            case NbtInt value:
                WriteInt32(value.Value);
                break;
            case NbtLong value:
                WriteInt64(value.Value);
                break;
            case NbtFloat value:
                // NaN や -0.0 を保つため、ビットパターンをそのまま書く
                WriteInt32(BitConverter.SingleToInt32Bits(value.Value));
                break;
            case NbtDouble value:
                WriteInt64(BitConverter.DoubleToInt64Bits(value.Value));
                break;
            case NbtByteArray value:
                WriteByteArrayPayload(value.Value);
                break;
            case NbtString value:
                WriteString(value.Value);
                break;
            case NbtList value:
                WriteListPayload(value);
                break;
            case NbtCompound value:
                WriteCompoundPayload(value);
                break;
            case NbtIntArray value:
                WriteIntArrayPayload(value.Value);
                break;
            case NbtLongArray value:
                WriteLongArrayPayload(value.Value);
                break;
            default:
                throw SpringNbtException.Malformed($"書き出せないタグ: {tag.Type.AsString()}");
        }
    }

    private void WriteCompoundPayload(NbtCompound compound)
    {
        // 挿入順のまま「タグID + 名前 + ペイロード」を並べる
        foreach (KeyValuePair<string, NbtTag> entry in compound)
        {
            buffer.Add((byte)entry.Value.Type);
            WriteString(entry.Key);
            WritePayload(entry.Value);
        }

        buffer.Add((byte)TagType.End);
    }

    private void WriteListPayload(NbtList list)
    {
        buffer.Add((byte)list.ElementType);
        WriteInt32(list.Count);

        // 要素型は共通なので、ペイロードだけを並べる
        foreach (NbtTag item in list)
        {
            WritePayload(item);
        }
    }

    private void WriteByteArrayPayload(sbyte[] values)
    {
        WriteInt32(values.Length);

        // 1 バイトずつそのまま書く
        foreach (sbyte value in values)
        {
            buffer.Add((byte)value);
        }
    }

    private void WriteIntArrayPayload(int[] values)
    {
        WriteInt32(values.Length);

        // 4 バイトずつビッグエンディアンで書く
        foreach (int value in values)
        {
            WriteInt32(value);
        }
    }

    private void WriteLongArrayPayload(long[] values)
    {
        WriteInt32(values.Length);

        // 8 バイトずつビッグエンディアンで書く
        foreach (long value in values)
        {
            WriteInt64(value);
        }
    }

    private void WriteString(string text)
    {
        byte[] encoded = Mutf8.Encode(text);

        // 長さフィールドは u16
        // NbtString 側でも検査しているが、キー名は素の string なのでここでも見る
        if (encoded.Length > Mutf8.MaxByteLength)
        {
            throw SpringNbtException.InvalidArgument(
                $"文字列が長すぎる: MUTF-8 で {encoded.Length} バイト (上限 {Mutf8.MaxByteLength})");
        }

        Span<byte> lengthBytes = stackalloc byte[2];
        BinaryPrimitives.WriteUInt16BigEndian(lengthBytes, (ushort)encoded.Length);
        buffer.Add(lengthBytes[0]);
        buffer.Add(lengthBytes[1]);
        buffer.AddRange(encoded);
    }

    private void WriteInt16(short value)
    {
        Span<byte> bytes = stackalloc byte[2];
        BinaryPrimitives.WriteInt16BigEndian(bytes, value);
        buffer.Add(bytes[0]);
        buffer.Add(bytes[1]);
    }

    private void WriteInt32(int value)
    {
        Span<byte> bytes = stackalloc byte[4];
        BinaryPrimitives.WriteInt32BigEndian(bytes, value);

        // Span は AddRange に渡せないので 1 バイトずつ積む
        for (int i = 0; i < bytes.Length; i++)
        {
            buffer.Add(bytes[i]);
        }
    }

    private void WriteInt64(long value)
    {
        Span<byte> bytes = stackalloc byte[8];
        BinaryPrimitives.WriteInt64BigEndian(bytes, value);

        // Span は AddRange に渡せないので 1 バイトずつ積む
        for (int i = 0; i < bytes.Length; i++)
        {
            buffer.Add(bytes[i]);
        }
    }
}

File: java/src/main/java/io/github/scriptarts/springnbt/nbt/NbtBinaryWriter.java
package io.github.scriptarts.springnbt.nbt;

import io.github.scriptarts.springnbt.SpringNbtException;
import java.io.ByteArrayOutputStream;
import java.util.Map;

/**
 * NBT を展開済みのバイト列へ書き出す
 *
 * <p>出力は一意でなければならない（ラウンドトリップ検証が成立するため）
 * Compound は挿入順のまま、浮動小数点はビットパターンのまま書き出す
 *
 * <p>仕様: {@code docs/spec/10-nbt-binary.md} 5章
 */
final class NbtBinaryWriter {

    private final ByteArrayOutputStream buffer = new ByteArrayOutputStream();

    /** ルートタグを書き出し、結果のバイト列を返す */
    byte[] writeRoot(NamedTag named, NbtFormat format) {
        buffer.write(TagType.COMPOUND.id());

        if (format == NbtFormat.JAVA) {
            // ファイル形式のルートには名前が付く
            writeString(named.name());
        }

        writeCompoundPayload(named.tag());
        return buffer.toByteArray();
    }

    private void writePayload(NbtTag tag) {
        switch (tag) {
            case NbtByte value -> buffer.write(value.value());
            case NbtShort value -> writeUnsigned(value.value(), 2);
            case NbtInt value -> writeUnsigned(value.value(), 4);
            case NbtLong value -> writeUnsigned(value.value(), 8);
            // NaN や -0.0 を保つため、ビットパターンをそのまま書く
            case NbtFloat value -> writeUnsigned(Float.floatToRawIntBits(value.value()), 4);
            case NbtDouble value -> writeUnsigned(Double.doubleToRawLongBits(value.value()), 8);
            case NbtByteArray value -> writeByteArrayPayload(value.value());
            case NbtString value -> writeString(value.value());
            case NbtList value -> writeListPayload(value);
            case NbtCompound value -> writeCompoundPayload(value);
            case NbtIntArray value -> writeIntArrayPayload(value.value());
            case NbtLongArray value -> writeLongArrayPayload(value.value());
        }
    }

    private void writeCompoundPayload(NbtCompound compound) {
        // 挿入順のまま「タグID + 名前 + ペイロード」を並べる
        for (Map.Entry<String, NbtTag> entry : compound) {
            buffer.write(entry.getValue().type().id());
            writeString(entry.getKey());
            writePayload(entry.getValue());
        }

        buffer.write(TagType.END.id());
    }

    private void writeListPayload(NbtList list) {
        buffer.write(list.elementType().id());
        writeUnsigned(list.size(), 4);

        // 要素型は共通なので、ペイロードだけを並べる
        for (NbtTag item : list) {
            writePayload(item);
        }
    }

    private void writeByteArrayPayload(byte[] values) {
        writeUnsigned(values.length, 4);
        buffer.writeBytes(values);
    }

    private void writeIntArrayPayload(int[] values) {
        writeUnsigned(values.length, 4);

        // 4 バイトずつビッグエンディアンで書く
        for (int value : values) {
            writeUnsigned(value, 4);
        }
    }

    private void writeLongArrayPayload(long[] values) {
        writeUnsigned(values.length, 4);

        // 8 バイトずつビッグエンディアンで書く
        for (long value : values) {
            writeUnsigned(value, 8);
        }
    }

    private void writeString(String text) {
        byte[] encoded = Mutf8.encode(text);

        // 長さフィールドは u16
        // キー名は素の String なのでここでも検査する
        if (encoded.length > Mutf8.MAX_BYTE_LENGTH) {
            throw SpringNbtException.invalidArgument(
                    "文字列が長すぎる: MUTF-8 で " + encoded.length + " バイト (上限 "
                            + Mutf8.MAX_BYTE_LENGTH + ")");
        }

        writeUnsigned(encoded.length, 2);
        buffer.writeBytes(encoded);
    }

    /** 値をビッグエンディアンで指定バイト数ぶん書く */
    private void writeUnsigned(long value, int count) {
        // 上位バイトから順に取り出す
        for (int index = count - 1; index >= 0; index--) {
            buffer.write((int) ((value >>> (index * 8)) & 0xFF));
        }
    }
}
